File: src/firebase/seedData.js
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
} from 'firebase/auth'
import {
  collection,
  doc,
  serverTimestamp,
  setDoc,
  addDoc,
} from 'firebase/firestore'
import { auth, db } from './config'

// Seed database with demo data.
// Call seedDatabase() once in a development-only context (e.g. from a temporary button).
export async function seedDatabase() {
  // Simple guard: avoid accidental reseeding in production.
  if (import.meta.env.PROD) {
    throw new Error('seedDatabase() should only be run in development.')
  }

  // Sign out any existing session to avoid conflicts
  try {
    await auth.signOut()
  } catch {
    // ignore
  }

  const createdUsers = {}

  async function createUser(email, password, role, extraUserData = {}) {
    const { user } = await createUserWithEmailAndPassword(auth, email, password)
    const userDoc = {
      name: extraUserData.name ?? email.split('@')[0],
      email,
      role,
      avatar: extraUserData.avatar ?? '👤',
      createdAt: serverTimestamp(),
    }
    await setDoc(doc(db, 'users', user.uid), userDoc)
    createdUsers[email] = { uid: user.uid, role }
    return user
  }

  try {
    // Create core users
    const admin = await createUser('admin@edukids.com', 'admin123', 'admin', {
      name: 'School Admin',
      avatar: '🏫',
    })

    const teacher1 = await createUser(
      'teacher1@edukids.com',
      'teacher123',
      'teacher',
      { name: 'Mr. Martin', avatar: '👨‍🏫' },
    )
    const teacher2 = await createUser(
      'teacher2@edukids.com',
      'teacher123',
      'teacher',
      { name: 'Ms. Dupont', avatar: '👩‍🏫' },
    )

    const parent1 = await createUser(
      'parent1@edukids.com',
      'parent123',
      'parent',
      { name: 'Parent One', avatar: '👨‍👩‍👧' },
    )
    const parent2 = await createUser(
      'parent2@edukids.com',
      'parent123',
      'parent',
      { name: 'Parent Two', avatar: '👨‍👩‍👧' },
    )
    const parent3 = await createUser(
      'parent3@edukids.com',
      'parent123',
      'parent',
      { name: 'Parent Three', avatar: '👨‍👩‍👧' },
    )

    const student1 = await createUser(
      'student1@edukids.com',
      'student123',
      'student',
      { name: 'Ahmed', avatar: '🧒' },
    )
    const student2 = await createUser(
      'student2@edukids.com',
      'student123',
      'student',
      { name: 'Sara', avatar: '👧' },
    )
    const student3 = await createUser(
      'student3@edukids.com',
      'student123',
      'student',
      { name: 'Léa', avatar: '👧' },
    )
    const student4 = await createUser(
      'student4@edukids.com',
      'student123',
      'student',
      { name: 'Noah', avatar: '🧒' },
    )

    // Link students to parents/teachers
    const studentsCol = collection(db, 'students')
    const studentDocs = []

    async function addStudent(studentUser, parentUser, teacherUser, classLevel) {
      const ref = await addDoc(studentsCol, {
        uid: studentUser.uid,
        name: studentUser.displayName || studentUser.email.split('@')[0],
        parentId: parentUser.uid,
        teacherId: teacherUser.uid,
        classLevel,
        xpPoints: 240,
        badges: ['Early Bird', 'Math Star'],
      })
      studentDocs.push({ id: ref.id, uid: studentUser.uid, classLevel })
      return ref
    }

    await addStudent(student1, parent1, teacher1, 'CM1')
    await addStudent(student2, parent1, teacher1, 'CM1')
    await addStudent(student3, parent2, teacher2, 'CE2')
    await addStudent(student4, parent3, teacher2, 'CE2')

    // Courses
    const coursesCol = collection(db, 'courses')
    const mathCourseRef = await addDoc(coursesCol, {
      title: 'Fun Maths',
      subject: 'Mathematics',
      teacherId: teacher1.uid,
      description: 'Numbers, games and puzzles',
      classLevel: 'CM1',
      videoUrl: 'https://www.youtube.com/watch?v=dQw4w9WgXcQ',
      createdAt: serverTimestamp(),
    })
    const frenchCourseRef = await addDoc(coursesCol, {
      title: 'French Stories',
      subject: 'French',
      teacherId: teacher2.uid,
      description: 'Reading, writing and storytelling',
      classLevel: 'CE2',
      videoUrl: '',
      createdAt: serverTimestamp(),
    })
    const scienceCourseRef = await addDoc(coursesCol, {
      title: 'Science Explorers',
      subject: 'Science',
      teacherId: teacher1.uid,
      description: 'Experiments and discoveries',
      classLevel: 'CM1',
      videoUrl: '',
      createdAt: serverTimestamp(),
    })

    // Assignments
    const assignmentsCol = collection(db, 'assignments')
    const hw1Ref = await addDoc(assignmentsCol, {
      courseId: mathCourseRef.id,
      title: 'Homework 1: Fractions',
      description: 'Practice adding and subtracting fractions.',
      dueDate: serverTimestamp(),
      type: 'homework',
    })
    const quiz1Ref = await addDoc(assignmentsCol, {
      courseId: mathCourseRef.id,
      title: 'Quiz 1: Multiplication',
      description: 'Short quiz on multiplication tables.',
      dueDate: serverTimestamp(),
      type: 'quiz',
    })
    const test1Ref = await addDoc(assignmentsCol, {
      courseId: scienceCourseRef.id,
      title: 'Test: The Solar System',
      description: 'End-of-unit test about the planets.',
      dueDate: serverTimestamp(),
      type: 'test',
    })
    const hw2Ref = await addDoc(assignmentsCol, {
      courseId: frenchCourseRef.id,
      title: 'Homework: Reading',
      description: 'Read a short story and answer questions.',
      dueDate: serverTimestamp(),
      type: 'homework',
    })
    const quiz2Ref = await addDoc(assignmentsCol, {
      courseId: frenchCourseRef.id,
      title: 'Quiz: Vocabulary',
      description: 'Match words with their meanings.',
      dueDate: serverTimestamp(),
      type: 'quiz',
    })

    // Grades
    const gradesCol = collection(db, 'grades')
    async function addGrade(studentUser, assignmentRef, courseRef, score, max, feedback) {
      await addDoc(gradesCol, {
        studentId: studentUser.uid,
        assignmentId: assignmentRef.id,
        courseId: courseRef.id,
        score,
        maxScore: max,
        feedback,
        gradedAt: serverTimestamp(),
      })
    }

    await addGrade(student1, hw1Ref, mathCourseRef, 18, 20, 'Great work!')
    await addGrade(student2, hw1Ref, mathCourseRef, 15, 20, 'Good effort, review fractions.')
    await addGrade(student1, quiz1Ref, mathCourseRef, 9, 10, 'Excellent!')
    await addGrade(student3, hw2Ref, frenchCourseRef, 16, 20, 'Nice reading.')
    await addGrade(student4, quiz2Ref, frenchCourseRef, 8, 10, 'Very good.')

    // Attendance (last 2 weeks)
    const attendanceCol = collection(db, 'attendance')
    const today = new Date()
    for (let i = 0; i < 14; i += 1) {
      const d = new Date(today)
      d.setDate(today.getDate() - i)

      const statusOptions = ['present', 'absent', 'late']
      const randomStatus = (indexOffset) =>
        statusOptions[(i + indexOffset) % statusOptions.length]

      for (const student of [student1, student2, student3, student4]) {
        await addDoc(attendanceCol, {
          studentId: student.uid,
          date: d,
          status: randomStatus(student.uid.length),
          markedBy: student.uid === student1.uid || student.uid === student2.uid ? teacher1.uid : teacher2.uid,
        })
      }
    }

    // Messages between teacher and parent
    const messagesCol = collection(db, 'messages')
    await addDoc(messagesCol, {
      senderId: teacher1.uid,
      receiverId: parent1.uid,
      senderRole: 'teacher',
      content: 'Ahmed has been doing great in maths this week!',
      read: false,
      sentAt: serverTimestamp(),
    })
    await addDoc(messagesCol, {
      senderId: parent1.uid,
      receiverId: teacher1.uid,
      senderRole: 'parent',
      content: "Thank you! We're very proud of him.",
      read: false,
      sentAt: serverTimestamp(),
    })

    console.info('EduKids seed data created successfully.')

    // Re-auth as admin for convenience
    await signInWithEmailAndPassword(auth, 'admin@edukids.com', 'admin123')

    return createdUsers
  } catch (error) {
    console.error('Error seeding EduKids data:', error)
    throw error
  }
}

File: backend/app/seed_data.py
"""Seed logic for EduKids. Used by seed_db.py and /api/seed."""
from datetime import datetime, timedelta
from app.models import User, Student, Course, Assignment, Grade, Attendance, Message
from app.auth import get_password_hash


def run_seed(db):
    # Users
    admin = User(
        email="admin@edukids.com",
        hashed_password=get_password_hash("admin123"),
        name="School Admin",
        role="admin",
        avatar="🏫",
        approved=True,
    )
    db.add(admin)
    db.flush()

    t1 = User(
        email="teacher1@edukids.com",
        hashed_password=get_password_hash("teacher123"),
        name="Mr. Martin",
        role="teacher",
        avatar="👨‍🏫",
        approved=True,
    )
    db.add(t1)
    db.flush()

    t2 = User(
        email="teacher2@edukids.com",
        hashed_password=get_password_hash("teacher123"),
        name="Ms. Dupont",
        role="teacher",
        avatar="👩‍🏫",
        approved=True,
    )
    db.add(t2)
    db.flush()

    p1 = User(
        email="parent1@edukids.com",
        hashed_password=get_password_hash("parent123"),
        name="Parent One",
        role="parent",
        avatar="👨‍👩‍👧",
        approved=True,
    )
    db.add(p1)
    db.flush()

    p2 = User(
        email="parent2@edukids.com",
        hashed_password=get_password_hash("parent123"),
        name="Parent Two",
        role="parent",
        avatar="👨‍👩‍👧",
        approved=True,
    )
    db.add(p2)
    db.flush()

    s1 = User(
        email="student1@edukids.com",
        hashed_password=get_password_hash("student123"),
        name="Ahmed",
        role="student",
        avatar="🧒",
        approved=True,
    )
    db.add(s1)
    db.flush()

    s2 = User(
        email="student2@edukids.com",
        hashed_password=get_password_hash("student123"),
        name="Sara",
        role="student",
        avatar="👧",
        approved=True,
    )
    db.add(s2)
    db.flush()

    s3 = User(
        email="student3@edukids.com",
        hashed_password=get_password_hash("student123"),
        name="Léa",
        role="student",
        avatar="👧",
        approved=True,
    )
    db.add(s3)
    db.flush()

    s4 = User(
        email="student4@edukids.com",
        hashed_password=get_password_hash("student123"),
        name="Noah",
        role="student",
        avatar="🧒",
        approved=True,
    )
    db.add(s4)
    db.flush()

    for user, parent, teacher, level in [
        (s1, p1, t1, "CM1"),
        (s2, p1, t1, "CM1"),
        (s3, p2, t2, "CE2"),
        (s4, p2, t2, "CE2"),
    ]:
        db.add(
            Student(
                user_id=user.id,
                parent_id=parent.id,
                teacher_id=teacher.id,
                class_level=level,
                xp_points=240,
                badges='["Early Bird", "Math Star"]',
            )
        )

    math_course = Course(
        title="Fun Maths",
        subject="Mathematics",
        teacher_id=t1.id,
        description="Numbers, games and puzzles",
        class_level="CM1",
        video_url="https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    )
    db.add(math_course)
    db.flush()

    french_course = Course(
        title="French Stories",
        subject="French",
        teacher_id=t2.id,
        description="Reading, writing and storytelling",
        class_level="CE2",
        video_url="",
    )
    db.add(french_course)
    db.flush()

    science_course = Course(
        title="Science Explorers",
        subject="Science",
        teacher_id=t1.id,
        description="Experiments and discoveries",
        class_level="CM1",
        video_url="",
    )
    db.add(science_course)
    db.flush()

    due = datetime.utcnow() + timedelta(days=7)
    hw1 = Assignment(
        course_id=math_course.id,
        title="Homework 1: Fractions",
        description="Practice adding and subtracting fractions.",
        due_date=due,
        type="homework",
        max_score=20,
    )
    db.add(hw1)
    db.flush()

    quiz1 = Assignment(
        course_id=math_course.id,
        title="Quiz 1: Multiplication",
        description="Short quiz on multiplication tables.",
        due_date=due,
        type="quiz",
        max_score=10,
    )
    db.add(quiz1)
    db.flush()

    hw2 = Assignment(
        course_id=french_course.id,
        title="Homework: Reading",
        description="Read a short story and answer questions.",
        due_date=due,
        type="homework",
        max_score=20,
    )
    db.add(hw2)
    db.flush()

    db.add(Grade(student_id=s1.id, assignment_id=hw1.id, course_id=math_course.id, score=18, max_score=20, feedback="Great work!"))
    db.add(Grade(student_id=s2.id, assignment_id=hw1.id, course_id=math_course.id, score=15, max_score=20, feedback="Good effort."))
    db.add(Grade(student_id=s1.id, assignment_id=quiz1.id, course_id=math_course.id, score=9, max_score=10, feedback="Excellent!"))
    db.add(Grade(student_id=s3.id, assignment_id=hw2.id, course_id=french_course.id, score=16, max_score=20, feedback="Nice reading."))

    today = datetime.utcnow().date()
    for i in range(5):
        d = datetime(today.year, today.month, today.day) - timedelta(days=i)
        for student in [s1, s2, s3, s4]:
            db.add(
                Attendance(
                    student_id=student.id,
                    date=d,
                    status="present" if i % 3 != 0 else "late",
                    marked_by=t1.id if student.id in (s1.id, s2.id) else t2.id,
                )
            )

    db.add(
        Message(
            sender_id=t1.id,
            receiver_id=p1.id,
            sender_role="teacher",
            content="Ahmed has been doing great in maths this week!",
            read=False,
        )
    )
    db.add(
        Message(
            sender_id=p1.id,
            receiver_id=t1.id,
            sender_role="parent",
            content="Thank you! We're very proud of him.",
            read=False,
        )
    )

    db.commit()
